### PR TITLE
feat(tabs): PF1:1

### PR DESCRIPTION
**Pull request by zeroedin**
## What I did
- removed all scss styling 
- implemented Patternfly v4 styles and API
- improved keyboard navigation, alignment with [W3C aria authoring practices for tabs](https://www.w3.org/WAI/ARIA/apg/example-index/tabs/tabs-automatic.html)

### TODO:
- [x] break out base classes for inheritance
- [x] implement overflow styles 
- [x] implement inset styles 
- [x] implement icon tab styles
- [x] implement filled styles
- [x] fix spec tests
- [x] improve docs (add CSS props jsdoc comments)
- [x] add changeset

### Choices:
for now, does not implement 
- horizontal overflow dropdown feature
- sub tabs feature
- nav element feature
- separate content (trigger) feature
- child tab-panel mounting features
- dynamic closable tabs feature
- loading a tab via an external toggle

## Testing Instructions
- Load up the deploy preview for this branch
- Load up [PatternFly v4 Tabs](https://www.patternfly.org/v4/components/tabs/)
- Compare and contrast
- Read over the [docs](https://deploy-preview-2099--patternfly-elements.netlify.app/components/tabs/) and see if anything's missing or no longer relevant

## Notes to reviewers
This PR is *BREAKING*. It aligns pfe-tabs with pfv4.

### CODE REVIEW

**Comment by bennypowers** [comment]
Should support icons in tabs per https://github.com/patternfly/patternfly-elements/discussions/2127#discussion-4335469

**Comment by zeroedin** [comment]
Demo is broken until #2137 is resolved

**Comment by zeroedin** [comment]
**Warning** this was forced pushed

**Comment by coreyvickery** [comment]
@zeroedin I noticed some weirdness with the overflow tabs, here is a recording.

https://drive.google.com/file/d/1DpAGcXg8-Q32Ow3jKzmZ4fJqzZruzwyL/view?usp=sharing

Everything else looks good.

**Comment by zeroedin** [comment]
> @zeroedin I noticed some weirdness with the overflow tabs, here is a recording.

Nice find @coreyvickery, turns out just a wrong CSS property mistake during a find & replacement was the culprit.  Fix pushed.

**Review by coreyvickery — APPROVED** [review]

File: elements/pfe-tabs/demo/pfe-tabs.js
Comment (bennypowers) [inline elements/pfe-tabs/demo/pfe-tabs.js:13]: is `active-key` how they do it in pfv4? i'd prefer `active-index` or even better, just put `active` on the active tab pair, but expose `.activeIndex`. WDYT?

File: elements/pfe-tabs/pfe-tab-panel.ts
Comment (bennypowers) [inline elements/pfe-tabs/pfe-tab-panel.ts:18]: does this need to be async?

File: elements/pfe-tabs/pfe-tab.scss
Comment (bennypowers) [inline elements/pfe-tabs/pfe-tab.scss:2]:
```suggestion
  margin-inline-end: var(--pf-c-tabs__link--child--MarginRight, var(--pf-global--spacer--md, 1rem));
```

File: elements/pfe-tabs/pfe-tab.ts
Comment (bennypowers) [inline elements/pfe-tabs/pfe-tab.ts:27]: as elsewhere, maybe we can use firstUpdated to avoid async connectedCallback. this is a small nit

**Review by bennypowers — CHANGES_REQUESTED** [review]
This is a tremendous amount of work, really nicely done. Most comments here are minor nits. Looking forward to releasing this.

File: elements/pfe-tabs/demo/pfe-tabs.html
Comment (zeroedin) [reply elements/pfe-tabs/demo/pfe-tabs.html:23]: This was to reproduce the [Patternfly v4 Tabs Aria Disabled (Tooltip)](https://www.patternfly.org/v4/components/tabs) example. 

> "Aria-disabled tabs are like disabled tabs, but focusable. Allows for tooltip support."   

However, our tooltip does not currently work in tabs due to the nature of how it is positioned on the page in the context of the element it is labeling.  Patternfly's is absolutely positioned and injected at the bottom of the page.

File: elements/pfe-tabs/demo/pfe-tabs.html
Comment (zeroedin) [reply elements/pfe-tabs/demo/pfe-tabs.html:4]: Will remove

File: elements/pfe-tabs/demo/pfe-tabs.js
Comment (zeroedin) [reply elements/pfe-tabs/demo/pfe-tabs.js:55]: Do we need to support less than version 13.1? https://caniuse.com/?search=ResizeObserver

File: elements/pfe-tabs/pfe-tab-panel.ts
Comment (zeroedin) [reply elements/pfe-tabs/pfe-tab-panel.ts:18]: No it doesn't, moving `this.#upgradeAccessibility()` to firstUpdated().

File: elements/pfe-tabs/BaseTabPanel.ts
Comment (zeroedin) [reply elements/pfe-tabs/BaseTabPanel.ts:23]: 🔥 that would be a great update

File: elements/pfe-tabs/BaseTabs.ts
Comment (zeroedin) [reply elements/pfe-tabs/BaseTabs.ts:21]: That actually just needs to be removed, it was orphaned.

File: elements/pfe-tabs/BaseTabs.ts
Comment (zeroedin) [reply elements/pfe-tabs/BaseTabs.ts:49]: 👍

File: elements/pfe-tabs/demo/pfe-tabs.js
Comment (bennypowers) [reply elements/pfe-tabs/demo/pfe-tabs.js:13]: i mean that PfeTabs.prototype.activeIndex should have a getter/setter pair, and update the element as expected, but it shouldn't reflect to or read from an attr, rather, users should just set `active` on the tab they want.

File: elements/pfe-tabs/demo/pfe-tabs.js
Comment (zeroedin) [reply elements/pfe-tabs/demo/pfe-tabs.js:7]: removed

File: elements/pfe-tabs/demo/pfe-tabs.js
Comment (bennypowers) [reply elements/pfe-tabs/demo/pfe-tabs.js:55]: lgtm. nice!

File: elements/pfe-tabs/BaseTabs.ts
Comment (zeroedin) [reply elements/pfe-tabs/BaseTabs.ts:66]: cleaned that up and added a disconnectedCallback to remove the window listener.

File: elements/pfe-tabs/BaseTabs.ts
Comment (zeroedin) [reply elements/pfe-tabs/BaseTabs.ts:309]: fixed, should have all been private.

File: elements/pfe-tabs/pfe-tab.ts
Comment (zeroedin) [reply elements/pfe-tabs/pfe-tab.ts:27]: 👍

File: elements/pfe-tabs/BaseTabs.ts
Comment (zeroedin) [reply elements/pfe-tabs/BaseTabs.ts:98]: I think that is probably a good idea.  Taking a stab.  Might have a question or two on the approach.

File: elements/pfe-tabs/demo/pfe-tabs.js
Comment (zeroedin) [reply elements/pfe-tabs/demo/pfe-tabs.js:11]: Ok gotcha, will explore that!

File: elements/pfe-tabs/BaseTabs.scss
Comment (zeroedin) [reply elements/pfe-tabs/BaseTabs.scss:176]: I guess it just depends if we want the scroll bar displayed or not.  This is the style defined by pfv4 for the overflow.   My personal preference is to leave the scrollbar visible for usability.  If you agree, I think it's safe to remove it. My assumption hiding it is also probably an accessibility concern as well.

File: elements/pfe-tabs/demo/pfe-tabs.html
Comment (zeroedin) [reply elements/pfe-tabs/demo/pfe-tabs.html:79]: We can reevaluate later to see if there is need to be able to turn off the scrollbuttons.

File: elements/pfe-tabs/BaseTabs.ts
Comment (bennypowers) [inline elements/pfe-tabs/BaseTabs.ts:219]: does this need to be protected?
if the subclass needs to hook in here, let's expose an optional protected hook instead
```ts
  #onSlotChange(): void {
    this.allTabs = this._tabs;
    this.#firstLastClasses();
    this.allPanels = this._panels;
    this.#updateAccessibility();
    this.onSlotchange?.();
  }
  
  protected onSlotchange?(): void
```

(notice also the [lower `c`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLSlotElement/slotchange_event))

File: elements/pfe-tabs/BaseTabs.ts
Comment (bennypowers) [inline elements/pfe-tabs/BaseTabs.ts:250]:
```suggestion
  #onTabExpand(event: Event): void {
```

### this binding
doesn't need the decorator, since `addEventListener` is called on the element instance (L59). same for `_onKeyDownHandler`. Similarly, anything bound in a render template listener will get auto-bound to the element instance, so `@bouunud _scrollLeft` can be `#scrollLeft`, etc
for handleScrollButtons, the listener on tablist should probably be removed in disconnectedcallback. We could remove the dependency on `@bound` by making it a class field. I'm not yet solid on which of those is really preferable, but my feeling now is that if that's the last instanced of `@bound` in the file, then better to remove it.
```ts
#onScroll = () => {
  clearTimeout(...)
  ...
}
```

### Methods Naming
let's use `onEvent` instead of `eventHandler` - it's fewer bytes and more consistent

**Review by bennypowers — CHANGES_REQUESTED** [review]

File: elements/pfe-tabs/demo/pfe-tabs.html
Comment (bennypowers) [inline elements/pfe-tabs/demo/pfe-tabs.html:182]: @hellogreg take a look at this particular case.

The internal plumbing is dependent on index, so the nth button always corresponds to the nth panel, regardless of pre-existing relationship in the markup.

So in your opinion should we
1. Be ok with this and merge - we don't need to support this case 
2. Merge now in order to unblock development but open a new issue to support this case
3. Block merging until this case is supported

**Review by bennypowers — APPROVED** [review]
I'm good, pending @hellogreg 's review

Well wanna go back and update the tests after switch lands, tho

**Comment by zeroedin** [comment]
> Well wanna go back and update the tests after switch lands, tho

Alrighty then,  I will also have to rework what I just committed, assuming switch has element internals stuff in it then.   I knew I was missing something as I was struggling to get tests/code to work properly.  All good can revert it later.

**Comment by bennypowers** [comment]
i can't shake the suspicion that we're bumping up on polyfill bugs

https://discord.com/channels/944894512717770762/945061812351692830/1037871477778829403

**Comment by bennypowers** [comment]
I think this works fine in chrome if we just set the relevant properties on `#internals`

https://codepen.io/bennyp/pen/BaVKQja?editors=1011

if it's not working in polyfill browsers, we should fix the polyfill, I'm certain our PR will be reviewed in short order. 

Likewise if FF's impl is broken we should file a bug

Screenshot from 2022-11-04 11-17-54

File: elements/pfe-tabs/demo/pfe-tabs.html
Comment (hellogreg) [reply elements/pfe-tabs/demo/pfe-tabs.html:182]: After talking with Steven about this, I say Let's Grab Two Martinis for now, but maybe revisit later.

**Review by hellogreg — APPROVED** [review]

**Pull request merged**
